### PR TITLE
Filter tasks earlier

### PR DESCRIPTION
**Pull request by jvegreg**
I am working on (hopefully) finalishing ESMValGroup/ESMValTool#1657 and realised that for the diagnostics filter to work you need to have the data for all the diagnostics that are not going to be run. That is a big problem, specially when using the `recipe_check_obs.yml` to check your data.

This PR fixes this by skipping creation of all tasks that are not meant to run.

Important clarification from @jvegreg ->

@valeriupredoi, the whole point of this change is to avoid creating the preprocessor tasks, as the execution is currently failing if some of the non required preprocessors does not have the data available. It is also a performance issue, as even if it not fails, we are spending the time to look for it and as you know that's non negligible

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by schlunma** [comment]
Just tested with `esmvaltool run --diagnostics diag_ecs  ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml` but got the following error:

```
2021-08-05 11:25:40,705 UTC [2641] ERROR   Program terminated abnormally, see stack trace below for more information:
Traceback (most recent call last):
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_main.py", line 432, in run
    fire.Fire(ESMValTool())
  File "/mnt/lustre02/work/bd0854/b309141/miniconda3/envs/esm/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/mnt/lustre02/work/bd0854/b309141/miniconda3/envs/esm/lib/python3.9/site-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/mnt/lustre02/work/bd0854/b309141/miniconda3/envs/esm/lib/python3.9/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_main.py", line 409, in run
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_main.py", line 100, in process_recipe
    recipe = read_recipe_file(recipe_file, config_user)
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_recipe.py", line 61, in read_recipe_file
    return Recipe(raw_recipe,
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_recipe.py", line 1017, in __init__
    self.tasks = self.initialize_tasks() if initialize_tasks else None
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_recipe.py", line 1439, in initialize_tasks
    self._resolve_diagnostic_ancestors(tasks)
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_recipe.py", line 1368, in _resolve_diagnostic_ancestors
    raise RecipeError(
esmvalcore._recipe_checks.RecipeError: Could not find any ancestors matching diag_psi/tas
```

[main_log.txt](https://github.com/ESMValGroup/ESMValCore/files/6938199/main_log.txt)
[main_log_debug.txt](https://github.com/ESMValGroup/ESMValCore/files/6938197/main_log_debug.txt)

One more thing: I think Codecov is not working properly: You added multiple lines to `_recipe.py`, but it doesn't show this in the report...

**Comment by jvegasbsc** [comment]
> Just tested with esmvaltool run --diagnostics diag_ecs ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml but got the following error:

Thanks @schlunma, I will take a look. Those ancestores should not be checked at all.

**Comment by jvegasbsc** [comment]
Ups, I was filtering preprocessor tasks and not diagnostic tasks

**Comment by jvegasbsc** [comment]
By the way, nice recipe to check:

- A diagnostic without preprocessed data
- A normal diagnostic
- A diagnostic without data that depends on the previous two

It's the better recipe I can think of to test this PR, also because it involves not so much data

**Comment by schlunma** [comment]
I tested again with the following commands:

These ran fine:

```
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'wrong_diag' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_ecs' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_psi' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_psi diag_ecs' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_wrong diag_ecs' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_cox18nature' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
```

This one failed:

```
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_cox18nature' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml

> 2021-08-11 10:58:58,667 UTC [16232] ERROR   Program terminated abnormally, see stack trace below for more information:
Traceback (most recent call last):
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_main.py", line 432, in run
    fire.Fire(ESMValTool())
  File "/mnt/lustre02/work/bd0854/b309141/miniconda3/envs/esm/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/mnt/lustre02/work/bd0854/b309141/miniconda3/envs/esm/lib/python3.9/site-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/mnt/lustre02/work/bd0854/b309141/miniconda3/envs/esm/lib/python3.9/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_main.py", line 409, in run
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_main.py", line 100, in process_recipe
    recipe = read_recipe_file(recipe_file, config_user)
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_recipe.py", line 61, in read_recipe_file
    return Recipe(raw_recipe,
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_recipe.py", line 1017, in __init__
    self.tasks = self.initialize_tasks() if initialize_tasks else None
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_recipe.py", line 1443, in initialize_tasks
    self._resolve_diagnostic_ancestors(tasks)
  File "/mnt/lustre01/pf/b/b309141/ESMValCore/esmvalcore/_recipe.py", line 1369, in _resolve_diagnostic_ancestors
    raise RecipeError(
esmvalcore._recipe_checks.RecipeError: Could not find any ancestors matching diag_ecs/ecs
```

I guess there is something wrong with ancestor handling.

[main_log_debug.txt](https://github.com/ESMValGroup/ESMValCore/files/6967884/main_log_debug.txt)

**Comment by jvegasbsc** [comment]
The case that is failing is due to a missing ancestor. I am not sure if we want to support this case:

- It has an easy workaround: `esmvaltool run --diagnostics 'diag_cox18nature diag_ecs/ecs diag_psi/psi'`
- It will make the code more complex and difficult to maintain, as we will need to check at task creation not only if a given task is in the filter but also if it is a prerequisite of one of the selected ones.

I don't know if this case is common enough to justify the extra complexity. What do you think, @ESMValGroup/esmvaltool-coreteam?

**Comment by bouweandela** [comment]
A quick grep through the existing recipes shows that about 31 of the 91 recipes use custom ancestors, that's about 1 in 3 so I think it's probably worth the effort.

**Comment by valeriupredoi** [comment]
I am very confused by this whole thing, and here are my two main confusion points:

- that recipe is a good example of how **not to** run just a single diagnostic out of a group of diagnostics - `diag_ecs` is an ancestor (pre-requisite) of the main diagnostic (the one that supplies the requested datasets and variables - how running just that one doesn't fail is beyond me. We should be very careful how we make that `--diagnostics` flag work and not just apply a technical approach to it;
- furthermore, if you want to have a diagnostic not run *and* the preprocessor associated with it not run (I think this is your aim right? no data finding means no preprocessor task from the diagnostic), change the bit in `_recipe.py`:
```python
        # Select only requested tasks
        tasks = tasks.flatten()
        if not run_diagnostic:
            tasks = TaskSet(t for t in tasks
                            if isinstance(t, PreprocessingTask))
```
`run_diagnostics` will be `False` for the diag instance that doesn't need to run, but what you want to do is to turn off its preprocessor tasks too

**Comment by jvegasbsc** [comment]
I will do it then.

@valeriupredoi, the whole point of this change is to avoid creating the preprocessor tasks, as the execution is currently failing if some of the non required preprocessors does not have the data available. It is also a performance issue, as even if it not fails, we are spending the time to look for it and as you know that's non negligible

**Comment by bouweandela** [comment]
This feature would also be useful as a simple way to run part of a recipe in combination with the `--resume_from` option for distributed computations #1128.

**Comment by bouweandela** [comment]
Coverage could a bit better, have the examples provided by @schlunma been implemented as unit tests already?

**Comment by schlunma** [comment]
Testing this again right now, will also try to fix the merge conflict.

**Comment by schlunma** [comment]
Fixing the merge conflict was really non-trivial, so it would be good if you have a second look at it, @jvegasbsc.

I tested all the use cases I can think of (see below), and all of them run fine now! :tada:  Thanks @jvegasbsc!

```
esmvaltool run --config-file ~/config/mymodels.yml ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'wrong_diag' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_ecs' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_psi' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_psi/tasa' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_psi diag_ecs' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_wrong diag_ecs' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
esmvaltool run --config-file ~/config/mymodels.yml --diagnostics 'diag_cox18nature' ~/ESMValTool/esmvaltool/recipes/recipe_cox18nature.yml
```

**Comment by schlunma** [comment]
I will fix the merge conflicts and add some tests so we can get this into v2.5.

**Comment by schlunma** [comment]
I fixed the merge conflicts and added extensive tests for the feature. I also re-tested every use case listed above. Everything works fine :tada:

I'm happy with this PR now, but since I added a lot of stuff myself it might be good if anyone else can take a look at it. Cheers!

**Comment by valeriupredoi** [comment]
cheers Manu! I'll have a look at it in a jiffy!

**Comment by valeriupredoi** [comment]
could you pls clean up the PR description/boxes/etc in the meantime :beer:

File: esmvalcore/_recipe.py [hunk 1]
```diff
@@ -1552,12 +1622,26 @@ def _fill_wildcards(self, variable_group, preprocessor_output):
...
+                                   tasknames_to_run, any_diag_script_is_run):
         """Create preprocessor tasks."""
         tasks = []
         failed_tasks = []
         for variable_group in diagnostic['preprocessor_output']:
             task_name = diagnostic_name + TASKSEP + variable_group
+
+            # Skip preprocessor if not a single diagnostic script is run and
+            # the preprocessing task is not explicitly requested by the user
+            if tasknames_to_run:
```
Comment (valeriupredoi) [inline esmvalcore/_recipe.py:1635]: why wouldn't `tasknames_to_run` be empty in this case? ie no `preprocessor` and no `diagnostics`

**Review by valeriupredoi — APPROVED** [review]
looking good! I have not tested it with actual runs, but that's why the nicely added tests are there for :grin: One thing I think we can extend this to is the special case of a `dry-run` when the user wants to run through the recipe without running any preprocessor (listed or not) and any diagnostic (again, listed or not) but just run the implicit preprocessing tasks (data finder, cmor checker and extra checks imposed by the preprocessor if any eg target grid  smaller than original grid, requested levels are within data lavels etc) - I reckon no task would run in this configuration of the code, but we'd have to add special tasks for those types of runs - I think that would be fine just by adding a new class of task, eg dry-run-task. Anyways, that's for the future :beer:

**Comment by valeriupredoi** [comment]
@jvegreg and @schlunma :beers: to you guys!

File: esmvalcore/_recipe.py (hunk 1 above)
Comment (schlunma) [reply esmvalcore/_recipe.py:1635]: Empty `tasknames_to_run` is the default value used when `--diagnostics` is not used; in this case, all tasks are run. But I agree, that feels weird, but much simpler to code.

Comment (valeriupredoi) [reply esmvalcore/_recipe.py:1635]: okies :+1:

**Review by schlunma — APPROVED** [review]

**Pull request merged**